### PR TITLE
V. cholerae O1/O139 mlst scheme selection bugfix

### LINKED ISSUE #58: Incorrect MLST scheme used for some Vibrio cholerae
_Vibrio cholerae_ has two MLST schemes, one for O1 or O139 serogroups, and one for non-O1/non-O139 serogroups. The latter is always being used regardless of serogroup. This needs to be amended

### PR DESCRIPTION
**Pull request by kapsakcj**
This PR closes #58 

🗑️ This dev branch should be deleted after merging to main.

## :brain: Aim, Context and Functionality

Vibrio cholerae has two MLST schemes, one for O1 or O139 serogroups, and one for non-O1/non-O139 serogroups. The latter is always being used regardless of serogroup, because `mlst`'s software auto scheme detection feature seems to always choose the non-O1 & non-O139 scheme. `mlst` also has a few default schemes that are excluded by default and `vcholerae_2` scheme is one of them, so perhaps that is why `vcholerae` is default.

The `mlst` command-line tool calls these schemes as such:

- Non-O1 & Non-O139 serogroup samples : `vcholerae`
- O1 & O139 serogroup samples, `vcholerae_2`

By default, the `mlst` auto-scheme detection chooses the non-O1 & O139 scheme so this is a way of automatically forcing `mlst` to use `vcholerae_2` for samples detected as O1 or O139 via the SRST2 task.

More info on the 2 mlst schemes can be found here: https://pubmlst.org/organisms/vibrio-cholerae

## :hammer_and_wrench:  Impacted Workflows/Tasks & Changes Being Made
This will affect the behavior of the workflow(s) even if users don’t change any workflow inputs relative to the last version : **No, but the exception is if they are analyzing V. cholerae samples that are O1 or O139 serogroups. The mlst results will change if analyzed w/ a previous version of TheiaProk wf**

Running this workflow on different occasions could result in different results, e.g. due to use of a live database, "latest" docker image, or stochastic data processing : **No**

#### 🔄 Data Processing 

Docker/software or software versions changed: **No**

Databases or database versions changed: **No**

Data processing/commands changed: **No**

File processing changed: **No**

Compute resources changed: **No**

#### ➡️ Inputs 

- `ts_mlst.scheme` has been exposed as an optional input for the TheiaProk_Illumina_PE workflow

## :microscope: Final Developer Checklist

- [x] The workflow/task has been tested locally and results, including file contents, are as anticipated
- [x] The workflow/task has been tested on Terra and results, including file contents, are as anticipated
- [x] The CI/CD has been adjusted and tests are passing (to be completed by Theiagen developer)
- [x] Code changes follow the [style guide](https://theiagen.notion.site/Style-Guide-WDL-Workflow-Development-bb456f34322d4f4db699d4029050481c)

## 🎯 Reviewer Checklist 

- [ ] All impacted workflows/tasks have been tested on Terra with a different dataset than used for development
- [ ] All reviewer-suggested scenarios have been tested and any additional
- [ ] All changed results have been confirmed to be accurate
- [ ] All workflows/tasks impacted by change/s have been tested using a standard validation dataset to ensure no unintended change of functionality
- [ ] All code adheres to the style guide
- [ ] MD5 sums have been updated
- [ ] The PR author has addressed all comments

## 🗂️ Associated Documentation (to be completed by Theiagen developer)

- [x] Relevant documentation on the Public Health Resources "PHB Main" has been updated. **No need to update docs unless we want to add a caveat/note about the auto-scheme selection for Vcholerae O1 and O139 samples**
- [x] Workflow diagrams have been updated to reflect changes. **No need to update wf diagram**

### CODE REVIEW

**Comment by kapsakcj** [comment]
TODO: need to update theiaprok ILMN SE, ONT, and FASTA workflows to get mlst scheme input from merlin_magic subworkflow.

Testing that this works for ILMN PE in Terra now

File: workflows/utilities/wf_merlin_magic.wdl [hunk 1]
```diff
@@ -366,6 +367,10 @@ workflow merlin_magic {
           srst2_min_edge_depth = srst2_min_edge_depth,
           srst2_gene_max_mismatch = srst2_gene_max_mismatch
       }
+    # this is nested inside of conditional check for merlin_tag so should only be available for Vibrio samples (meaning srst2 task was run)
```
Comment (kapsakcj) [inline workflows/utilities/wf_merlin_magic.wdl:370]: and now I'm considering moving it inside of the conditional `if (!assembly_only && !ont_data)`...

File: workflows/utilities/wf_merlin_magic.wdl [hunk 2]
```diff
@@ -675,6 +680,8 @@ workflow merlin_magic {
     String? srst2_vibrio_toxR = srst2_vibrio.srst2_vibrio_toxR
     String? srst2_vibrio_serogroup = srst2_vibrio.srst2_vibrio_serogroup
     String? srst2_vibrio_biotype = srst2_vibrio.srst2_vibrio_biotype
+    # this output mainly for vibrio, but also for non-vibrio samples. prioritize vibrio output; but otherwise pass along user-defined input, or use 2 single quotes (trick mlst software into using auto-scheme detection feature)
+    String ts_mlst_scheme_out = select_first([ts_mlst_scheme_vibrio, ts_mlst_scheme, "''"])
```
Comment (kapsakcj) [inline workflows/utilities/wf_merlin_magic.wdl:684]: change order so ts_mlst_scheme comes first (user-defined)

followed by vibrio one

followed by 2 single quotes

File: workflows/theiaprok/wf_theiaprok_illumina_pe.wdl [hunk 3]
```diff
@@ -59,6 +59,7 @@ workflow theiaprok_illumina_pe {
     String genome_annotation = "prokka" # options: "prokka" or "bakta"
     String? expected_taxon  # allow user to provide organism (e.g. "Clostridioides_difficile") string to amrfinder. Useful when gambit does not predict the correct species    # qc check parameters
     File? qc_check_table
+    String? ts_mlst_scheme
```
Comment (kapsakcj) [inline workflows/theiaprok/wf_theiaprok_illumina_pe.wdl:62]: change to: `mlst_scheme`

File: workflows/theiaprok/wf_theiaprok_illumina_pe.wdl [hunk 4]
```diff
@@ -230,8 +226,15 @@ workflow theiaprok_illumina_pe {
...
           read1 = read_QC_trim.read1_clean,
-          read2 = read_QC_trim.read2_clean
+          read2 = read_QC_trim.read2_clean,
+          ts_mlst_scheme = ts_mlst_scheme
       }
+        call ts_mlst_task.ts_mlst {
+          input: 
+            assembly = shovill_pe.assembly_fasta,
+            samplename = samplename,
+            scheme = merlin_magic.ts_mlst_scheme_out
```
Comment (kapsakcj) [inline workflows/theiaprok/wf_theiaprok_illumina_pe.wdl:236]: change to `merlin_magic.ts_mlst_scheme`

File: workflows/utilities/wf_merlin_magic.wdl (hunk 1 above)
Comment (kapsakcj) [reply workflows/utilities/wf_merlin_magic.wdl:370]: I cannot read, the check is already nested inside of !assembly_only and !ont_data conditional 

Just going to update the comment to clarify

**Comment by kapsakcj** [comment]
Tests launched & comments regarding them:

- TheiaProk_Illumina_PE; PHE Vibrio cholerae dataset, mix of O1's, O139's, and non O1s & O139s; did not provide any optional input for `mlst_scheme`: https://app.terra.bio/#workspaces/theiagen-validations/curtis-sandbox-theiagen-validations/job_history/464f6fef-f9f0-4a36-929b-265cf8c2567f
  - ran as expected, used `vcholerae_2` scheme for those designated as O1/O139 from SRST2 task and used `vcholerae` for all other Vcholerae samples
- TheiaProk_Illumina_PE; A. baumanii samples; did not provide `mlst_scheme` as input: https://app.terra.bio/#workspaces/theiagen-validations/curtis-sandbox-theiagen-validations/job_history/17e64f6a-97b5-441f-9a0e-57d4ac764066
  - ran as expected, and used the 2 single quotes `"''"` as input for `mlst_scheme` to trigger the `mlst` scheme auto-selection feature
- TheiaProk_Illumina_PE; 2 Shigella samples; DID provide `mlst_scheme` as input (used `"ecoli_achtman_4"`): https://app.terra.bio/#workspaces/theiagen-validations/curtis-sandbox-theiagen-validations/job_history/942e9f52-4eaf-4446-9b0c-34f76fae8c25
  - ran as expected, used the higher-priority `mlst_scheme` provided input

TODO:

- [ ] functional test of TheiaProk_FASTA (at least 1 sample, does not need to be Vcholerae)
- [ ] functional test of TheiaProk_ONT (at least 1 sample, does not need to be Vcholerae)
- [ ] functional test of TheiaProk_Illumina_SE (at least 1 sample, does not need to be Vcholerae)

**Comment by kapsakcj** [comment]
One last test to show that user-defined input is priority in the select_first statement in merlin_magic subworkflow.

I used `"senterica"` as the input mlst_scheme variable for a few E. coli and the mlst task used `senterica` as the scheme: 

https://app.terra.bio/#workspaces/theiagen-validations/curtis-sandbox-theiagen-validations/job_history/ddd76e8c-fc7e-467d-a3e5-307cec43020c

**Comment by emmadoughty** [comment]
I have looked deeper into the literature given Curtis' observation that no ST was assigned to the O1 and O139 samples with the `vcholerae_2` scheme. 

image

- Whilst PubMLST lists the Kanampalliwar & Singh scheme (`vcholerae_2`)published in 2020 as the O1 and O139 scheme, it doesn't seem to be used widely (only 5 citations to the paper) and only[ 400 PubMLST submissions to the scheme ](https://pubmlst.org/bigsdb?db=pubmlst_vcholerae_seqdef&page=schemeInfo&scheme_id=2) have ever been made, all on one date, presumably when the scheme was first published. The [paper publishing the scheme](https://pubmed.ncbi.nlm.nih.gov/32982995/). described that "there is no proper MLST database for typing V. cholerae O1 and O139 strains, except for non-O1, non-O139 strains (Octavia et al., 2013)." hence them developing their scheme.
- The _Octavia et al._ scheme (`vcholerae`) published in 2013 is listed as the non-O1, non-O139 on PubMLST and the title of the [paper publishing this scheme](https://pubmed.ncbi.nlm.nih.gov/23776471/) mentions only these serotypes, non-O1 and non-O139 . On reading the paper, the scheme was developed based on housekeeping genes identified in [previous work](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3367346/) looking at more diverse _V. cholerae_ (including O1). _Octavia et al._ had published the MLST scheme in a paper that was also describing non-O1 and non-O139 isolates, but I can't see any references to this MLST scheme only being for non-O1/non-O139 other than on PubMLST and in the Kanampalliwar & Singh paper. The _Octavia et al._ scheme has been used widely, with 85 citations on Google scholar. These seem to mostly be people using the MLST scheme, and they use it for O1 and O139. It also has 1800 submissions to PubMLST. 

Though PubMLST lists the `vcholerae_2` scheme as the O1 and O139 scheme, it may be more useful for the public health community to use the `vcholerae` scheme as they are more likely to identify a named ST for their _V. cholerae_, including O1 and O139 isolates, and the use of this scheme doesn't seem to be wrong- in fact, there seems to be a consensus to preferentially using this.

**Comment by cimendes** [comment]
Thank you @emmadoughty for the detective work 🕵🏻‍♀️ 

I agree with your assessment that the community is using `vcholerae` scheme for all isolates, regardless of serotype. I think we should close off this PR and not incorporate these changes into the main codebase to keep in line with the community standards and expectations.

**Pull request closed without merging**